File: 刷题/剑指offer/二叉树/重建二叉树.cpp
class Solution {
public:
    TreeNode* R(<vector<int> a, int abegin,int aend, vector<int> b,int bbegin,int bend)
    {
        if(abegin>=aend || bbegin>=bend)
            return NULL;
        TreeNode* root=new TreeNode(a[abegin]);
        int pivot;  //在中序序列中找到root所在的位置
        for(pivot=bbegin;pivot<bend;pivot++)
            if(b[pivot]==a[abegin])
                break;
        root->left=R(a,abegin+1,abegin+pivot-bbegin+1,b,bbegin,pivot);
        //left: [abegin+1, abegin+pivot-bbegin+1)  最后的加1，是取不到的
        //      [bbegin, pivot)
        root->right=R(a,abegin+pivot-bbegin+1,aend,b,pivot,bend);
        //right: [abegin+pivot-bbegin+1,aend)
        //       [pivot, bend)
        return root;
        
    }


    TreeNode* reConstructBinaryTree(vector<int> pre,vector<int> vin) {
        return R(pre,0,pre.size(),vin,0,vin.size());  //因为起始从0开始，所以是一个左闭右开的区间
    }
};

File: 刷题/剑指offer/二叉树/重建二叉树.py
# python
class Solution:
    def reConstructBinaryTree(self, pre, vin):
        if not pre or not vin:
            return None
        root = TreeNode(pre.pop(0))  # 根结点
        index = vin.index(root.val)  # 中序中根结点的位置
        root.left = self.reConstructBinaryTree(pre, vin[:index])
        root.right = self.reConstructBinaryTree(pre, vin[index + 1:])
        return root
